src/dialects/abstract/data-types.ts

```diff
@@ -0,0 +1,1826 @@
+  public static readonly key = 'TEXT';
+  protected readonly options: TextOptions;
+  protected _length?: TextLength;
+
+  /**
+   * @param [length='TEXT'] could be tiny, medium, long.
+   */
+  constructor(length?: TextLength | Partial<TextOptions>) {
+    super();
+
+    const options = typeof length === 'object' ? length : { length };
+
+    this.options = {
+      length: options.length?.toLowerCase() as TextLength,
+    };
+
+    this._length = this.options.length;
+  }
+
+  toSql() {
+    switch (this.options.length) {
+      case 'tiny':
+        return 'TINY_TEXT';
+      case 'medium':
+        return 'MEDIUM_TEXT';
+      case 'long':
+        return 'LONG_TEXT';
+      default:
+        return this.key;
+    }
+  }
+
+  validate(value: string): asserts value is string;
+  validate(value: string): true {
+    if (typeof value !== 'string') {
+      throw new ValidationError(
+        util.format('%j is not a valid string', value),
+        [],
+      );
+    }
+
+    return true;
+  }
+}
+
+/**
+ * Unlimited length TEXT column
+ */
+export type TEXT = _TEXT;
+export const TEXT = classToInvokable(_TEXT);
+
+class _CITEXT extends ABSTRACT<string> {
+  public static readonly key = 'CITEXT';
+  validate(value: string): asserts value is string;
+  validate(value: string): true {
+    if (typeof value !== 'string') {
+      throw new ValidationError(
+        util.format('%j is not a valid string', value),
+        [],
+      );
+    }
+
+    return true;
+  }
+}
+
+/**
+ * An unlimited length case-insensitive text column.
+ * Original case is preserved but acts case-insensitive when comparing values (such as when finding or unique constraints).
+ * Only available in Postgres and SQLite.
+ *
+ */
+export type CITEXT = _CITEXT;
+export const CITEXT = classToInvokable(_CITEXT);
+
+export interface NumberOptions {
+  length?: number;
+  decimals?: number;
+  zerofill?: boolean;
+  unsigned?: boolean;
+}
+
+type AcceptedNumber =
+  | number
+  | boolean
+  | null
+  | undefined
+  | boolean
+  | { toString(): string };
+
+class _NUMBER extends ABSTRACT<AcceptedNumber, number> {
+  public static readonly key: string = 'NUMBER';
+
+  protected options: NumberOptions;
+  protected _length?: number;
+  protected _decimals?: number;
+  protected _zerofill?: boolean;
+  protected _unsigned?: boolean;
+
+  /**
+   * @param options type options
+   * @param [options.length] length of type, like `INT(4)`
+   * @param [options.zerofill] Is zero filled?
+   * @param [options.unsigned] Is unsigned?
+   * @param [options.decimals] number of decimal points, used with length `FLOAT(5, 4)`
+   * @param [options.precision] defines precision for decimal type
+   * @param [options.scale] defines scale for decimal type
+   */
+  constructor(options: number | Readonly<NumberOptions> = {}) {
+    super();
+    if (typeof options === 'number') {
+      options = {
+        length: options,
+      };
+    }
+
+    this.options = options;
```

Comment:
I think it would be a good opportunity to normalize `options`. We should either have `options`, or `_length` & co. But not both, otherwise we need to keep both in sync and it's error prone.